server/src/main/java/org/apache/druid/segment/realtime/appenderator/BatchAppenderator.java

```diff
@@ -0,0 +1,1335 @@
+          Execs.newBlockingSingleThreaded("[" + StringUtils.encodeForFormat(myId) + "]-appenderator-abandon", 0)
+      );
+    }
+  }
+
+  private void shutdownExecutors()
+  {
+    if (persistExecutor != null) {
+      persistExecutor.shutdownNow();
+    }
+
+    if (pushExecutor != null) {
+      pushExecutor.shutdownNow();
+    }
+
+    if (intermediateTempExecutor != null) {
+      intermediateTempExecutor.shutdownNow();
+    }
+  }
+
+  @VisibleForTesting
+  public List<File> getPersistedidentifierPaths()
+  {
+
+    ArrayList<File> retVal = new ArrayList<>();
+
+    final File baseDir = tuningConfig.getBasePersistDirectory();
+    if (!baseDir.exists()) {
+      return null;
+    }
+
+    final File[] files = baseDir.listFiles();
+    if (files == null) {
+      return null;
+    }
+
+    for (File sinkDir : files) {
+      final File identifierFile = new File(sinkDir, IDENTIFIER_FILE_NAME);
+      if (!identifierFile.isFile()) {
+        // No identifier in this sinkDir; it must not actually be a sink directory. Skip it.
+        continue;
+      }
+      retVal.add(sinkDir);
+    }
+
+    return retVal;
+  }
+
+  Pair<SegmentIdWithShardSpec, Sink> getIdentifierAndSinkForPersistedFile(File identifierPath)
+  {
+
+    try {
+      final SegmentIdWithShardSpec identifier = objectMapper.readValue(
+          new File(identifierPath, "identifier.json"),
+          SegmentIdWithShardSpec.class
+      );
+
+      // To avoid reading and listing of "merged" dir and other special files
+      final File[] sinkFiles = identifierPath.listFiles(
+          (dir, fileName) -> !(Ints.tryParse(fileName) == null)
+      );
+
+      Arrays.sort(
+          sinkFiles,
+          (o1, o2) -> Ints.compare(Integer.parseInt(o1.getName()), Integer.parseInt(o2.getName()))
+      );
+
+      List<FireHydrant> hydrants = new ArrayList<>();
+      for (File hydrantDir : sinkFiles) {
+        final int hydrantNumber = Integer.parseInt(hydrantDir.getName());
+
+        log.debug("Loading previously persisted partial segment at [%s]", hydrantDir);
+        if (hydrantNumber != hydrants.size()) {
+          throw new ISE("Missing hydrant [%,d] in identifier [%s].", hydrants.size(), identifier);
+        }
+
+        hydrants.add(
+            new FireHydrant(
+                new QueryableIndexSegment(indexIO.loadIndex(hydrantDir), identifier.asSegmentId()),
+                hydrantNumber
+            )
+        );
+      }
+
+      Sink currSink = new Sink(
+          identifier.getInterval(),
+          schema,
+          identifier.getShardSpec(),
+          identifier.getVersion(),
+          tuningConfig.getAppendableIndexSpec(),
+          tuningConfig.getMaxRowsInMemory(),
+          maxBytesTuningConfig,
+          null,
+          hydrants
+      );
+      currSink.finishWriting(); // this sink is not writable
+      return new Pair<>(identifier, currSink);
+    }
+    catch (IOException e) {
```

Comment:
Every exception should be propagated properly. Please rethrow it. Then this method doesn't have to return null.

Reply:
Done